### DISCUSSION TITLE
Is there a way to find out where the caret index is in TextEdit or LineEdit?

### DISCUSSION THREAD

**Discussion by maxenko** [slint-ui/slint#5123] [Q&A] [answered]
Is there any way to set it?

**Comment by syf20020816** [comment]
It seems we just can find what change in textedit or lineedit

maybe you need this callback in TextInput(define textedit yourself):

cursor-position-changed(Point): The cursor was moved to the new (x, y) position described by the [Point](https://releases.slint.dev/1.5.1/docs/slint/src/language/builtins/structs#point) argument.

**Reply by maxenko** [reply]
Thank you, I am trying to wrap my mind around how I can translate x,y of the point to the _caret index_ within the string of text of the editor. Or what the use of having pixel x,y would be in general when editing text...?

This is markup editing situation, and user would use keyboard shortcuts to modify words based on where the caret is, so detecting text edit delta and figuring out where the change is, is not an option.

This is something that is normally part of bare bones essentials in text editing widgets/controls. Surprised there isn't just:
**in-out property caret-index**

**Reply by syf20020816** [reply]
It seems you need to use callback edited, It always be called after input text change
`edited()`: Invoked when the text has changed because the user modified it.
So you can make a private prop to record the pre input text and dif with the lastest input text, and then you can get what people change
---
Use x,y is a challenge. It is hard to use x,y to get what people edit. Even single char has different width, so you need calc a lot (you can see the width if you use Text widget). Perhaps English may be easier because it can be combined with 26 letters. But a large number of punctuation marks may be more cumbersome.
Base on these, (x,y) is not a great way

**Reply by maxenko** [reply]
pixel x,y is completely useless, because it varies entirely on font and possibly even screen scale. It sounds like a simple case of not exposing some internal property.

Just to be clear with an example: in this github markdown editor, I want to make word 'clear' earlier in the sentence bold, I put my caret in the middle of the word and click [B], which makes it bold (double *): **clear**.

That is all I am looking for.

**Reply by syf20020816** [reply]
Fortunately, I have provided a component that can perceive user input（but you need to do some further processing, as I said add a private prop to record the last text）：https://surrealism-all.github.io/SurrealismUI.github.io/components/form/input.html

**Reply by syf20020816** [reply]
you can directly use the my source code and define yourself: https://github.com/Surrealism-All/SurrealismUI/blob/main/src/input/input.slint

**Reply by maxenko** [reply]
> private prop to record the last text

Text does not change. Only caret position does.

**Reply by syf20020816** [reply]
your desire can not be work in current support widget, It seems you need a widget like makedown editor or code editor, current text input can not support

**Reply by maxenko** [reply]
It could if it had caret index. This is a very simple request. Nothing fancy needed like a specialized editor, just an index within text where the caret is.

**Reply by syf20020816** [reply]
Yes, but this requires a more powerful editing method. When we need to bold certain text, users wrap it by adding * *, but in reality, the output result seen by the end user is already bold text, which reminds me of the real-time rendering of typroa. Of course, we can provide a basic editor for users to input, but currently real-time rendering seems impossible because widgets and HTML are different. Therefore, if you want to implement similar functions, you need an editing area and an edited result rendering area, which are two different widgets

**Reply by maxenko** [reply]
I don't need to show bold output. I just need to wrap the word with **

**Reply by syf20020816** [reply]
If that's all, what you really need most is lexical analysis

**Reply by maxenko** [reply]
what i really need is caret index, i'll take it from there.
I already do this in WPF, its mundane.

**Comment by syf20020816** [comment]
you may find some help in api doc https://releases.slint.dev/1.5.1/docs/rust/slint/struct.sharedstring

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by maxenko** [comment, marked as answer]
Ok found it. This is not documented unless you dig through the source (or pay more attention to intellisense, as I should have).

TextInput has:
```self.cursor-position-byte-offset```
```self.anchor-position-byte-offset```

```slint
TextInput {
    text: "Hello World";
    
    cursor-position-changed(p) => {
        debug(self.cursor-position-byte-offset);
    }
}
```
